### PR TITLE
Add a pass to vectorize tensor.pack and tensor.unpack ops.

### PR DESCRIPTION
**Pull request by hanhanW**
It adapts the LinalgExtVectorization pass to upstream tensor.pack and tensor.unpack version. It has the same test coverage comparing to LinalgExt version.

It also registers tensor tiling implementation to IREE becuase tiling tensor.pack and tensor.unpack op is needed for vectorization.

### CODE REVIEW

**Comment by hanhanW** [comment]
@chelini this is the vectorization pass for tensor.pack and tensor.unpack ops. Please take a look, thanks!

**Comment by hanhanW** [comment]
Register tiling interface for tensor dialect introduces infinite loop in `CPUDoubleTilingPadExpert`. 

Input of repro: https://gist.githubusercontent.com/hanhanW/9398a14b3f24359b8e56e6322d914245/raw

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      auto unpackTilingOptions =
+          scf::SCFTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder, Operation *op) {
+                Location loc = op->getLoc();
+                auto unpackOp = cast<tensor::UnPackOp>(op);
+                int numLoops = unpackOp.getDestRank();
+                auto dimAndTileMapping = unpackOp.getDimAndTileMapping();
+                SmallVector<Value> tileSizes;
+                for (int i = 0; i < numLoops; ++i) {
+                  if (dimAndTileMapping.count(i)) {
```
Comment (chelini) [inline compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:85]: Here I ended-up using tile sizes of 1 too as I did not find a way to generalize unpack this way when outer dims are present.

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      llvm::dbgs()
+          << "--- After generalizing tensor.pack and tensor.unpack ops ---\n";
+      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+      llvm::dbgs() << "\n\n";
+    });
+
+    // Kick in generic vectorizer.
+    {
+      RewritePatternSet patterns(ctx);
+      patterns.add<IREE::LinalgExt::LinalgVectorizationPattern>(ctx);
```
Comment (chelini) [inline compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:133]: Is all of this available in main too?

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-vectorize-pack-unpack-ops"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+/// A simple pattern rewriter that implements no special logic.
```
Comment (chelini) [inline compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:28]: There is something similar in `TransformUtils.h`.

**Review by chelini — COMMENTED** [review]
Nice, thanks!

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:133]: The implementation is all in MLIR repo. This patterns is just a wrapper for calling `vectorize` method.

https://github.com/iree-org/iree/blob/234f4b2f0f25d693e6e488908d5c5232851346af/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp#L258-L268

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp (hunk 3 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:28]: Thanks, good to know this. :)

File: compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp:80]: good point, thanks for pointing it out. I also updated the tests which cover the case!

**Comment by hanhanW** [comment]
This depends on https://reviews.llvm.org/D141977

File: compiler/src/iree/compiler/Codegen/Passes.h [hunk 4]
```diff
@@ -115,6 +115,8 @@ createConvertToDestinationPassingStylePass(
 /// control flows.
 std::unique_ptr<OperationPass<func::FuncOp>> createVectorizePadPass();
 
+std::unique_ptr<OperationPass<func::FuncOp>> createVectorizePackUnPackOpsPass();
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.h:118]: Please comment the pass.

**Review by MaheshRavishankar — APPROVED** [review]
Looks fine to me. Please wait for LLVM changes to propagate before landing these.

**Pull request merged**